odl/discr/discr_utils.py

```diff
@@ -0,0 +1,818 @@
+    >>> from odl.discr.grid import sparse_meshgrid
+    >>> mesh = sparse_meshgrid([0.0, 0.4, 1.0], [1.5, 3.5])
+    >>> interpolator(mesh)  # 3x2 grid of points
+    array([[ 2.,  3.],
+           [ 2.,  3.],
+           [ 6.,  7.]])
+
+    With nearest neighbor interpolation, we can also use non-scalar data
+    types like strings:
+
+    >>> part = odl.uniform_partition(0, 3, 6)
+    >>> part.coord_vectors  # grid points
+    (array([ 0.25,  0.75,  1.25,  1.75,  2.25,  2.75]),)
+    >>> f = ['s', 't', 'r', 'i', 'n', 'g']
+    >>> interpolator = nearest_interpolator(f, part.coord_vectors)
+    >>> print(interpolator(0.9))
+    t
+
+    See Also
+    --------
+    linear_interpolator : (bi-/tri-/...)linear interpolation
+    per_axis_interpolator : potentially different interpolation in each axis
+
+    Notes
+    -----
+    - **Important:** if called on a point array, the points are
+      assumed to be sorted in ascending order in each dimension
+      for efficiency reasons.
+    - Nearest neighbor interpolation is the only scheme which works
+      with data of non-numeric data type since it does not involve any
+      arithmetic operations on the values, in contrast to other
+      interpolation methods.
+    """
+    f = np.asarray(f)
+
+    # TODO(kohr-h): pass reasonable options on to the interpolator
+    def nearest_interp(x, out=None):
+        """Interpolating function with vectorization."""
+        x, x_type, x_is_scalar = _check_interp_input(x, f)
+        interpolator = _NearestInterpolator(coord_vecs, f, input_type=x_type)
+
+        res = interpolator(x, out=out)
+        if x_is_scalar:
+            res = res.item()
+        return res
+
+    return nearest_interp
+
+
+def linear_interpolator(f, coord_vecs):
+    """Return the linear interpolator for discrete values.
+
+    Parameters
+    ----------
+    f : numpy.ndarray
+        Function values that should be interpolated.
+    coord_vecs : sequence of numpy.ndarray
+        Coordinate vectors of the rectangular grid on which interpolation
+        should be based. They must be sorted in ascending order. Usually
+        they are obtained as ``grid.coord_vectors`` from a `RectGrid`.
+
+    Returns
+    -------
+    interpolator : function
+        Python function that will interpolate the given values when called
+        with a point or multiple points (vectorized).
+
+    Examples
+    --------
+    We interpolate a 1d function. If called with a single point, the
+    interpolator returns a single value, and with multiple points at once,
+    an array of values is returned:
+
+    >>> part = odl.uniform_partition(0, 2, 5)
+    >>> part.coord_vectors  # grid points
+    (array([ 0.2,  0.6,  1. ,  1.4,  1.8]),)
+    >>> f = [1.0, 2.0, 3.0, 4.0, 5.0]
+    >>> interpolator = linear_interpolator(f, part.coord_vectors)
+    >>> interpolator(0.3)  # 0.75 * 1 + 0.25 * 2 = 1.25
+    1.25
+    >>> interpolator([0.6, 1.3, 1.9])
+    array([ 2.  ,  3.75,  3.75])
```

Comment:
I still do not understand how the last component becomes `3.75`. The linear interpolation is between `(1.8, f(1.8)) = (1.8, 5.0)` and `(x, f(x))` for what `x` and what value of `f(x)`?

Since the partition is between `0` and `2`, I assume `x = 2`. But that would mean `f(2) = 2.5`, which I do not understand why...

**Edit:** This makes sense if `x = 2.2`, since the step size between points are `0.4`, and indeed we have zero boundary values, i.e., `f(2.2) = 0`. Is this correct? Then I do understand the examples below (line 429 and further) as well, with mixed linear and nearest interpolation.

Reply:
Yes, that's what happens. Sorry for not clarifying. Maybe I should write a comment into the example that explains the behavior.